jrt/src/com/yahoo/jrt/SslSocket.java

```diff
@@ -0,0 +1,232 @@
+
+    private static final Logger log = Logger.getLogger(SslSocket.class.getName());
+
+    private enum HandshakeState { NOT_STARTED, NEED_READ, NEED_WRITE, COMPLETED }
+
+    private final SocketChannel channel;
+    private final SSLEngine sslEngine;
+    private final ByteBuffer wrapBuffer;
+    private final ByteBuffer unwrapBuffer;
+    private final ByteBuffer handshakeDummyBuffer;
+    private HandshakeState handshakeState;
+
+    public SslSocket(SocketChannel channel, SSLEngine sslEngine) {
+        this.channel = channel;
+        this.sslEngine = sslEngine;
+        SSLSession nullSession = sslEngine.getSession();
+        this.wrapBuffer = ByteBuffer.allocate(nullSession.getPacketBufferSize() * 2);
+        this.unwrapBuffer = ByteBuffer.allocate(nullSession.getPacketBufferSize() * 2);
+        // Note: Dummy buffer as unwrap requires a full size application buffer even though no application data is unwrapped
+        this.handshakeDummyBuffer = ByteBuffer.allocate(nullSession.getApplicationBufferSize());
+        this.handshakeState = HandshakeState.NOT_STARTED;
+    }
+
+    @Override
+    public SocketChannel channel() {
+        return channel;
+    }
+
+    @Override
+    public HandshakeResult handshake() throws IOException {
+        HandshakeState newHandshakeState = processHandshakeState(this.handshakeState);
+        log.fine(() -> String.format("Handshake state '%s -> %s'", this.handshakeState, newHandshakeState));
+        this.handshakeState = newHandshakeState;
+        return toHandshakeResult(newHandshakeState);
+    }
+
+    private HandshakeState processHandshakeState(HandshakeState state) throws IOException {
+        switch (state) {
+            case NOT_STARTED:
+                sslEngine.beginHandshake();
+                break;
+            case NEED_WRITE:
+                channelWrite();
+                break;
+            case NEED_READ:
+                channelRead();
+                break;
+            case COMPLETED:
+                return HandshakeState.COMPLETED;
+            default:
+                throw new IllegalStateException("Unhandled state: " + state);
+        }
+
+        while (true) {
+            switch (sslEngine.getHandshakeStatus()) {
+                case NOT_HANDSHAKING:
+                    if (hasWrapBufferMoreData()) return HandshakeState.NEED_WRITE;
+                    sslEngine.setEnableSessionCreation(false); // disable renegotiation
+                    return HandshakeState.COMPLETED;
+                case NEED_TASK:
+                    sslEngine.getDelegatedTask().run();
+                    break;
+                case NEED_UNWRAP:
+                    if (hasWrapBufferMoreData()) return HandshakeState.NEED_WRITE;
+                    int bytesUnwrapped = sslEngineUnwrap(handshakeDummyBuffer);
+                    if (handshakeDummyBuffer.position() > 0) throw new SSLException("Got application data in handshake unwrap: " + handshakeDummyBuffer);
```

Comment:
Was pondering if this could trigger if a server peer pipelines its `Finished` packet with the first application data, but assuming unwrap() only processes 1 handshake frame per invocation during handshaking it _shouldn't_. Are my assumptions correct?

Reply:
Only a single frame is unwrapped at a time.